### ISSUE TITLE
submit stale share option?

### ISSUE THREAD

**Issue by miningpoolhub** [Genoil/cpp-ethereum#36]
How about adding submit stale share option?

Maybe I can do the work. Will you merge it to trunk when it's done?

**Comment by Genoil** [comment]
AFAIK the miner can't know whether or not a found share is stale. When a share is found, it is CPU-validated and then submitted. It's up to the pool whether or not to accept it. I've made a small change in 1.0.8 (unreleased because of some pending issues) where the CPU validation should produce less failures, esp. in getwork mode.

That said, if you mean something completely different, I'm open to suggestions, and contributions even more so!

**Comment by miningpoolhub** [comment]
I found that sometimes ethminer submits nonce and headerHash couple incorrectly. (nonce matches with previous headerHash but submitted with new headerHash resulting low difficulty share)

As I read the source code, I see only one headerHash variable from whole ethminer. Am I right?
current.headerHash contains only the latest data and not previous one.
Even below commit from 1.0.8, 
https://github.com/Genoil/cpp-ethereum/commit/def718be573512ec413d127aea7a16e26389cc00
Mutex wraps changing latest headerHash variable. it fixes nonce and headerHash mismatch issue but this is not an ideal solution regarding the stale shares.

For ethereum, stale shares can even make a block, ethminer should not rely only on latest one headerHash variable. I think headerHash value must be stored like a pool in memory and should be matched with correct nonce and submitted to pool.
As a pool operator, I added stale share building block feature few days ago and I found that stale shares finds new block or uncle blocks quite well. If ethminer handles multiple headerHash solutions correctly, mining efficiency can grow much better for pools like us.

**Comment by Genoil** [comment]
Yes there is only the current headerHash, both in getwork and stratum. I did not implement mutex in getWork yet, like I did in the Stratum client. Once I have this, what is then the goal of keeping old header hashes? This allows me to submit shares that could end up being stale pool-side, but as far as I understand that's the pool's responsibility, and apparently you've found a solution already.

**Comment by miningpoolhub** [comment]
The goal of keeping old header hash is to build new block by stale shares from pool.
I want ethminer to send all potential shares, because it "really" can make new block in ethereum world. This is not rubbish data. Some of them will be uncle, and some of them can be the main block also.

I think you read my suggestion opposite. I don't mean that pool don't want stale shares, but I wrote that pool want stale shares. How can a pool can get the share info when ethminer don't even send it. This cannot be solved from pool side.
Some GPU found solutions are lost by (1)nonce-headerHash mismatch at CPU verification and (2)new headerHash update from pool. Already found solutions became rubbish currently.

I did some research. Pool collected stale shares recently, and compared difficulty target to match, and when difficulty was higher than target, it could make several blocks successfully. Some of them became new block, and some of them became uncle block. These stale shares were sent from ethminer because it was not stale share when it was sent from client. (headerHash not updated at client side) About 1~3% new blocks could be generated.
I want to maximize this block finding potential by submitting stale shares from ethminer even new headerHash info is received from pool.

To summarize, two or three things must be solved to accomplish this.
(1) nonce, headerHash mistmatch problem fix
(2) multiple headerHash handling internally
(3) maybe --submit-stale-shares option add.

I hope I described the well and you get the point what I'm saying.

**Comment by Genoil** [comment]
I get the point. The goal of the 1.0.8 changes is to solve (1). The way I solved it will submit all shares found by ethminer regardless of the moment they potentially became invalid. The result is either Yay or Boo, and from what I understand your goal is to use the Boos to build uncles or even blocks. But I really don't see what I could do about it other than solve (1). Miner will only hash on one and only headerHash and solutions found will be validated with that header and sent with that. 

Perhaps @nerdralph can chip in because he's working on this, too

**Comment by nerdralph** [comment]
I did a quick hack on EthStratumClient::submit to send all shares even if they fail the EthashAux::eval.  I didn't go through the code in enough detail to realize the shares are submitted along with the headerhash.  So to do a proper test I need keep the current and last, and do an eval on both to tell if it is a stale share.

**Comment by Genoil** [comment]
but my point is that it doesn't matter whether the client knows if the share is stale or not. it's up to the pool to judge that.

**Comment by nerdralph** [comment]
But it does matter since it would add a lot more work on the pool side to check the current and last header hash.  It may be more clear if you look at the code (needs to be cleaned up a bit if it works) that I'm currently testing.
`
        if (EthashAux::eval(m_current.seedHash, m_current.headerHash, solution.nonce).value < m_current.boundary)
        {
                json = "{\"id\": 4, \"method\": \"mining.submit\", \"params\": [\"" + p_active->user + "\",\"" + m_job + "\",\"0x" + solution.nonce.hex() + "\",\"0x" + m_current.headerHash.hex() + "\",\"0x" + solution.mixHash.hex() + "\"]}\n";
        }
        else if (EthashAux::eval(m_last.seedHash, m_last.headerHash, solution.nonce).value < m_last.boundary)
        {
                cwarn << "submitting stale share";
                json = "{\"id\": 4, \"method\": \"mining.submit\", \"params\": [\"" + p_active->user + "\",\"" + m_job + "\",\"0x" + solution.nonce.hex() + "\",\"0x" + m_last.headerHash.hex() + "\",\"0x" + solution.mixHash.hex() + "\"]}\n";
        }
        else
        {
                cwarn << "FAILURE: GPU gave incorrect result!";
                return false;
        }
`

**Comment by nerdralph** [comment]
So far it is working perfectly.  526 solutions found, 520 matched the current hash and 6 matched the previous hash.  Not a single error (failure)!  This is a 1.2% improvement in mining.

**Comment by nerdralph** [comment]
It's been running for a couple more hours without problems, so I'd say its' ready to be merged.
Solutions found: 892
Sol. accepted  : 892
Stale shares   : 10

I tried creating a pull request, but since I already have a pull request for the exename, I can't seem to be able to create another one.  I guess I should have created another branch or tag first.  The changes are in my fork for anyone who wants to try it, or to merge into upstream.  I'm still learning git & github so it may be a day or two before I figure out how to tag or branch my two change sets.
https://github.com/nerdralph/cpp-ethereum

**Comment by Genoil** [comment]
Thanks I'm going to test run this of course !

**Comment by Genoil** [comment]
I've given it some thought and you guys are right. When there is new work, there is a small window where gpu threads could still call EthStratumClient::submit based on old work. 

Haven't seen any fails (0 out of 136 in 7.5hrs) in my last test run without your change though, but theoretically speaking it makes complete sense. So I'm implementing it and count the stales separately.

**Comment by nerdralph** [comment]
Try with global-work 16384 and 32768.  I tested with 32768.

**Comment by Genoil** [comment]
Ah that makes sense. Will postpone the test until later then. Can't use my PC with that amount of GPU work and I hate the noise my headless 7950 produces.

@miningpoolhub I will point the test at your server. Do you have a failover host?

**Comment by nerdralph** [comment]
With AMD and 15.7.1 drivers on Windows, using --global-work 32768 has a
minimal impact on system responsiveness, and 16384 makes no noticeable
impact.

**Comment by Genoil** [comment]
miningpoolhub share diff is pretty low so I get a relatively high number of shares for my single card setups :). So that's good for testing this. Still on 100% Accepted on current work though, but I'll leave it on for a night. Listed you in the credits as @nerdralph, let me know if you prefer  the other nick.

**Comment by nerdralph** [comment]
nerdralph is good; if people want to look at my github repo that's how
they'll find me.

**Comment by miningpoolhub** [comment]
@Genoil 

Here are lists of host you can use.

us-east1.ethereum.miningpoolhub.com
europe1.ethereum.miningpoolhub.com
asia1.ethereum.miningpoolhub.com
port 20535 is opened for Stratum mode.

MiningPoolHub uses vardiff. Difficulty will match to your hardware level as time passes.
If you want to start from default difficulty for immediate test, you'd better change workername since pool uses last worker's difficulty level to optimize.

**Comment by Genoil** [comment]
Just did a 2 hour run with "previous submit" about 27MH/s in total (970 + crappy 7950 with heat problems). Submitted 600+ shares, none of which were stales. My client says all were accepted, no rejects, no fails. So I wonder how much the change really does. Perhaps the mutex also fixes the issue. 
Pool says 2% invalid though. Have to double-check client reports ok.

**Comment by nerdralph** [comment]
I haven't looked through the mutex code, but that might well solve the stale share issue too.  But I don't think the mutex code is ready for deployment yet as I suspect a race condtion.

**Comment by davembg** [comment]
1.0.8 build works fine on Ubuntu 10.15 Cat 15.2 with (1)390x and (2)390 rig.
No deadlocks, been running for an hour:
ethminer -G -R /home/dave/dag --cl-local-work 256 --cl-global-work 16384 -v 9 --farm-recheck 400 -F http://www.donkeypool.com/0x45B187fF84d70f0677dF2b76eeB15F6453aef6d0/amd/95

60 shares, 95MH
donkeypool's standard diff is higher, so the shares have a much better chance of sealing the block.

As a side note - vardiff is a bit of a myth. It's not actually vardiff as implemented in other blockchains.
Also, your stratum client code doesn't actually set the difficulty, nor do pools that say they have vardiff.
They essentially adjust the 3rd param in the getwork RPC call to make miners think they are doing more work. You submit a lot of lower diff shares that don't have much chance of sealing the block. Because it's not implemented in Ethereum, it really doesn't have nearly 10-20% bump.

The main advantage to stratum is the blockchange response times - now if only the popular pools could agree on a standard implementation.

**Comment by miningpoolhub** [comment]
@Genoil 
As I monitored many shares from pool side, not all miners experience stale share thing frequently. As stale shares only happen when there's some gap between GPU and network, GPU hardware would affect much. If you want to test it, you would have to adjust GPU settings extremely or just input `sleep()` thing in source code to make deliberate time gap.

I read latest source code from 108 branch and found that `EthhashAux::eval()` and `std::ostream` were done in locked state in EthStratumClient::submit()
https://github.com/Genoil/cpp-ethereum/blob/2b7b4f625dd5ad17f47e25ba4f4bb5bdb64b99dd/libstratum/EthStratumClient.cpp#L378

This would make some problems.
(1) deadlock happens if `eval()` or `std::ostream` function make exception inside, which will make not to run `x_current.unlock()` code.
(2) cost quite big time by calculating hashing in synchronized state which will make delay getting new work

I would like to suggest making temporary copied variable and doing hash verification with that
```
bool EthStratumClient::submit(EthashProofOfWork::Solution solution) {
	x_current.lock();
	WorkPackage tempWork(m_current);
	string temp_job = m_job;
	WorkPackage tempPreviousWork(m_previous);
	string temp_previous_job = m_previousJob;
	x_current.unlock();

	cnote << "Solution found; Submitting to" << p_active->host << "...";
	cnote << "  Nonce:" << "0x"+solution.nonce.hex();
	//cnote << "  Mixhash:" << "0x" + solution.mixHash.hex();
	//cnote << "  Header-hash:" << "0x" + m_current.headerHash.hex();
	//cnote << "  Seedhash:" << "0x" + m_current.seedHash.hex();
	//cnote << "  Target: " << "0x" + h256(m_current.boundary).hex();
	//cnote << "  Ethash: " << "0x" + h256(EthashAux::eval(m_current.seedHash, m_current.headerHash, solution.nonce).value).hex();

	if (EthashAux::eval(tempWork.seedHash, tempWork.headerHash, solution.nonce).value < tempWork.boundary)
	{
		string json = "{\"id\": 4, \"method\": \"mining.submit\", \"params\": [\"" + p_active->user + "\",\"" + temp_job + "\",\"0x" + solution.nonce.hex() + "\",\"0x" + tempWork.headerHash.hex() + "\",\"0x" + solution.mixHash.hex() + "\"]}\n";
		std::ostream os(&m_requestBuffer);
		os << json;
		m_stale = false;
		async_write(m_socket, m_requestBuffer,
			boost::bind(&EthStratumClient::handleResponse, this,
			boost::asio::placeholders::error));
		return true;
	}
	else if (EthashAux::eval(tempPreviousWork.seedHash, tempPreviousWork.headerHash, solution.nonce).value < m_previous.boundary)
	{
		string json = "{\"id\": 4, \"method\": \"mining.submit\", \"params\": [\"" + p_active->user + "\",\"" + temp_previous_job + "\",\"0x" + solution.nonce.hex() + "\",\"0x" + tempPreviousWork.headerHash.hex() + "\",\"0x" + solution.mixHash.hex() + "\"]}\n";
		std::ostream os(&m_requestBuffer);
		os << json;
		m_stale = true;
		cwarn << "Submitting stale solution.";
		async_write(m_socket, m_requestBuffer,
			boost::bind(&EthStratumClient::handleResponse, this,
			boost::asio::placeholders::error));
		return true;
	}
	else {
		m_stale = false;
		cwarn << "FAILURE: GPU gave incorrect result!";
		p_farm->failedSolution();
	}
	
	return false;
}
```

And I would like to suggest
`x_stale`
to be removed. x_stale makes locks only when x_current makes lock. I don't think twice lock is needed. `m_stale`
doesn't rely on x_stale also, so it looks okay to remove x_stale variable and lock thing.

It would be better if I comment with pull request, but I haven't made compile environment for ethminer yet, afraid to commit which is not even tested for compiliation. Above source code is not tested, just typed.


@davembg
vardiff is not myth.
If you set static high diff value (which is quite higher for 1 gpu miner), then this small miner's mining status will be not measured correctly, and his reward would be much smaller than actual his contribution.
In opposite, if you set static low difficulty for 1 gpu miner, then 6-way gpu miner would have to insert x6 shares more in same time period. With proxy things in ethereum, there could be +20 rigs with 6-way gpu. 120 gpu would insert x120 shares than small miner. If small miner inserts one share per 20 seconds, then 120gpu miner inserts every share in 0.16 second. This is like DDoS. Server and client both will suffer useless high traffic.
Many pools which don't support vardiff handle this problem by making high diff port only or make several diff level ports. This is not simple problem when pool gets bigger.

Actually, vardiff is nothing to do with blockchain. Even bitcoin don't have vardiff thing in blockchain. It's just some protocol between pool and miner to verify/monitor whether if miners are doing their work. Shares are not only for finding block, but it's verification process to make fair distribution after block is found.

**Comment by davembg** [comment]
@miningpoolhub I should've used "coins" instead of block chains, agreed. All you are saying is true, and essentially seems to agree with my brief description. What you're NOT saying is anything about increasing payouts by 10-20%. If your pool produces more than it's average hashrate, then I guess I believe it. My point is the 10-20% payout increase is a myth. Not that it makes it more "fair".

**Comment by davembg** [comment]
@miningpoolhub More to the point, vardiff with ETH is more fair for HONEST miners. If it was supported in the wallet (blockchain, sry) this wouldn't be an issue for debate. As it is now, there doesn't seem to be standard for implementing vardiff or stratum in the miner, or the pool, which will continue to be an issue.

**Comment by miningpoolhub** [comment]
@davembg 
I didn't mention 10~20% payout increase payout in this thread. What are you referring about?

Vardiff is not something should supported by wallet. Even bitcoin don't have something like that. It's just miner program and pool's protocol as I already wrote. rpc getWork's Third param you mentioned is used for difficulty boundary in current eth-proxy pool miner and it's enough for vardiff or diff purpose. Proxy uses already implemented third param for it, because net difficulty is not needed for pool miner, and compatible with local mining.
Stratum like protocol is already implemented in genoil-ethminer. (Same as coinotron's stratum-mining-proxy) It's not just proxy like rpc relaying protocol, true stratum like protocol. This can be said to be de facto standard as there's no other stratum like protocol for ethereum currently.
I think this thread is not for this subject. Please contact contact@miningpoolhub.com if you want further discussion or tell us info.

**Comment by Genoil** [comment]
@miningpoolhub agreed that should reduce the chance of deadlock. Just ran two other tests with the mutex completely disabled against mph @ 18MH/s, different global work. On the first run with global work 2048 I had 1 accepted stale (validated from m_previous) and 147 accepts. On the second run with global work 32768 (to see if it would make a difference) I had over 200 accepts and no stales. Pool-side it showed similar invalid rates (0.8% and 0%).  

Also completely removed the x_stale indeed.

**Comment by miningpoolhub** [comment]
@Genoil 

I think it can't be successfully run without mutex with this structure, since there's no way to atomically replace m_current. boundary, headerHash, and seedHash group would have different group's value when thread race condition happens. To accomplish it, I think it must be pointed by pointer and pointer value should be exchanged atomically.
Anyway I'm satisfied with your works, happy to see any implementation you choose because you know the whole structure well. Even without mutex, it will make very little chance of race condition error because boundary and seedHash would not change frequently. Only headerHash matters much.

MiningPoolHub don't return "job not found" even it's stale share. It just return OK to miner but writes to DB as stale share. It's because some proxy programs try to resend stale shares when OK is not received. As it makes hiccup and lower the performance, pool just pretend it as accepted. But some miners wanted to see the actual stats of stale, low difficulty, duplicate shares, we write them all to db. This is why you get invalid rates from website but all OK from ethminer. I saw your previous comments and I was about to write about it but forgot to write at that time. My Bad. Sorry.

**Comment by Genoil** [comment]
closing this as this seems to be implemented and working correct. thanks for the help guys!
